File: algorithms/1_easy/L/03105_longest_strictly_increasing_or_strictly_decreasing_subarray/solution.py
class Solution:
    def longestMonotonicSubarray(self, nums: List[int]) -> int:
        longest = 1
        curr = 0
        n = len(nums)
        state = 0

        for i in range(1, n):
            if nums[i] > nums[i - 1]:
                if state != 1:
                    curr = 2
                    state = 1
                else:
                    curr += 1
            elif nums[i] < nums[i - 1]:
                if state != -1:
                    curr = 2
                    state = -1
                else:
                    curr += 1
            else:
                state = 0
                curr = 1

            longest = max(longest, curr)

        return longest

File: algorithms/1_easy/L/03105_longest_strictly_increasing_or_strictly_decreasing_subarray/solution.cpp
class Solution {
public:
    int longestMonotonicSubarray(vector<int>& nums) {
        int longest{1};
        int curr{0};
        const size_t len{nums.size()};
        int state{0};

        for (size_t i{1}; i < len; ++i) {
            if (nums[i] > nums[i - 1]) {
                if (state != 1) {
                    curr = 2;
                    state = 1;
                } else {
                    curr += 1;
                }
            } else if (nums[i] < nums[i - 1]) {
                if (state != -1) {
                    curr = 2;
                    state = -1;
                } else {
                    curr += 1;
                }
            } else {
                state = 0;
                curr = 1;
            }

            longest = std::max(longest, curr);
        }

        return longest;
    }
};
